### PR TITLE
Weather Feature

### PR DESCRIPTION
**Pull request by amycommits**
# Frontend Test Suite Implementation

## Overview
This PR implements

- a call to the Open Weather API
- stores the information in Redis
- adding information to a queue that expires every 45 minutes
- a list of users
- a modal with detailed information about the user's weather
- front and backend unit tests

#### UserCard Component
- Added tests for rendering user information correctly
- Implemented click event testing for modal interaction
- Added error state handling tests
- All 3 tests passing ✅

#### UserWeatherModal Component
- Added tests for modal rendering with weather details
- Implemented close event testing (both button and backdrop clicks)
- Added error state handling tests
- All 4 tests passing ✅

#### useWeather Composable
- Implemented tests in both TypeScript and JavaScript
- Added successful data fetching tests
- Implemented error handling tests
- All 4 tests passing (2 in TS, 2 in JS) ✅

### 3. Test Infrastructure
- Set up Vitest configuration with JSDOM environment
- Configured test setup file with necessary mocks
- Added proper TypeScript support
- Implemented proper axios mocking using `vi.spyOn`

### 4. Configuration Updates
- Fixed PostCSS configuration to use CommonJS syntax
- Updated package.json with test scripts
- Added necessary dev dependencies

## Test Coverage
- Total Tests: 11
- Test Files: 4
- Components Covered: 2
- Composables Covered: 1

## Test Results
All tests are passing with the following breakdown:
- UserCard: 3/3 tests passing
- UserWeatherModal: 4/4 tests passing
- useWeather (JS): 2/2 tests passing
- useWeather (TS): 2/2 tests passing

## Notes
- One expected Vue warning in UserWeatherModal test for null weatherInfo prop (intentional for error case testing)
- All tests include proper error handling scenarios
- Tests are written to be maintainable and readable

## How to Test
1. Run frontend tests:
   ```bash
   npm run test
   ```
2. Run backend tests:
   ```bash
   docker exec -it fullstack-challenge-app-1 bash -c "cd /var/www/html && php artisan test"
   ```

## Future Improvements
- Add more edge case tests
- Implement E2E tests
- Add test coverage reporting
- Add more component interaction tests
- Allow users to sign up/login and choose their own profile picture

## Notes

I added a column to the users table so that I could store a random profile picture in the database. I thought that it would look nice on the frontend
MainListMockup
/user-attachments/assets/7c2abbc3-3937-4e60-a39b-2252e080ae56)

### CODE REVIEW

File: api/app/Jobs/UpdateUserWeather.php [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+     * Execute the job.
+     */
+    public function handle(): void
+    {
+        try {
+            $response = Http::timeout(self::REQUEST_TIMEOUT)
+                ->get("https://api.openweathermap.org/data/2.5/weather", [
+                    'lat' => $this->user->latitude,
+                    'lon' => $this->user->longitude,
+                    'appid' => env('OPENWEATHER_API_KEY'),
```
Comment (sean-dickinson) [inline api/app/Jobs/UpdateUserWeather.php:41]: I know you're newer to Laravel, but you can use the `config` here since you made the config/weather.php file. So this would be `config('weather.api_key')`. The advantage of using config is that it can be cached and can enforce types in a way that the env can't.

File: api/app/Jobs/UpdateUserWeather.php [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    /**
+     * Execute the job.
+     */
+    public function handle(): void
+    {
+        try {
+            $response = Http::timeout(self::REQUEST_TIMEOUT)
+                ->get("https://api.openweathermap.org/data/2.5/weather", [
```
Comment (sean-dickinson) [inline api/app/Jobs/UpdateUserWeather.php:38]:
```suggestion
                ->get(config('weather.base_url'), [
```
You can use your config here as well

File: api/app/Jobs/UpdateUserWeather.php
Comment (sean-dickinson) [inline api/app/Jobs/UpdateUserWeather.php:1]: I like that you made this job fetch a single User's weather. It's ideal to keep jobs small and atomic. 👍🏻

File: api/app/Jobs/UpdateUserWeather.php [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+use Illuminate\Support\Facades\Http;
+use Illuminate\Support\Facades\Redis;
+use Illuminate\Support\Facades\Log;
+
+class UpdateUserWeather implements ShouldQueue
+{
+    use Dispatchable, InteractsWithQueue, Queueable, SerializesModels;
+
+    protected $user;
+    protected const REQUEST_TIMEOUT = 0.5; // 500ms timeout
```
Comment (sean-dickinson) [inline api/app/Jobs/UpdateUserWeather.php:20]: 👍🏻 good idea to add a timeout here

File: api/app/Models/User.php [hunk 4]
```diff
@@ -20,6 +20,11 @@ class User extends Authenticatable
         'email',
         'password',
     ];
+    
+    public function weatherInfo()
+    {
+        return $this->hasOne(WeatherInfo::class);
```
Comment (sean-dickinson) [inline api/app/Models/User.php:26]: It looks like you added this relationship , but ended up using Redis to store this data instead of the database. I don't see a migration to add this table so I think you can get rid of this method.

File: api/app/Services/WeatherService.php [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+    public function fetchWeatherForUsers()
+    {
+        $users = User::all();
+        $weatherData = [];
+
+        foreach ($users as $user) {
+            $cacheKey = self::REDIS_PREFIX . $user->id;
+            
+            try {
```
Comment (sean-dickinson) [inline api/app/Services/WeatherService.php:32]: I like that you implemented a "stale-while-revalidate" pattern here to serve stale data so as not to slow down the response

File: api/database/seeders/DatabaseSeeder.php [hunk 6]
```diff
@@ -12,8 +13,34 @@ class DatabaseSeeder extends Seeder
      *
      * @return void
      */
-    public function run()
+    public function run(): void
     {
-         User::factory(20)->create();
+        // Create test users with coordinates
```
Comment (sean-dickinson) [inline api/database/seeders/DatabaseSeeder.php:18]: I think you probably could leave the original factory that seeds 20 users and add your own after if you want to have some more deterministic data in there.

File: api/env_for_cursor.txt
Comment (sean-dickinson) [inline api/env_for_cursor.txt:1]: Whoops I think you probably didn't mean to commit this file. Something like this should be .gitignored since it is your .env file which may have secrets in it (like the api key). However since you commited it I see you did make some changes to the .env, you should instead update the .env.example with the new keys that are needed to run the app.

File: api/tests/Feature/WeatherControllerTest.php [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+                    'weatherInfo' => [
+                        'main' => ['temp'],
+                        'weather' => [['description', 'icon']],
+                    ],
+                    'last_updated',
+                ],
+            ]);
+    }
+
+    public function test_get_users_weather_handles_no_data()
```
Comment (sean-dickinson) [inline api/tests/Feature/WeatherControllerTest.php:63]: In this test you could make the `Queue` facade so that you can assert that the jobs you expect are put on the queue.

**Review by sean-dickinson — COMMENTED** [review]
Thanks so much for taking the time to do our challenge! From what I hear this is your first real foray into Laravel so I want to congratulate you on taking that on as part of the challenge. Hopefully your familiarity with Rails helped with some of the core concepts at least.

I think you used Cursor to help out with this task, I'm interested in how you approached using it for something like this. Feel free to drop a comment on here with any context about that or anything else from this challenge you want us to know.

File: api/app/Jobs/UpdateUserWeather.php [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+            if ($response->successful()) {
+                $weatherInfo = $response->json();
+                $userWeatherInfo = [
+                    'name' => $this->user->name,
+                    'icon' => $this->user->profile_picture,
+                    'weatherInfo' => $weatherInfo,
+                    'last_updated' => now()->timestamp,
+                ];
+
+                Redis::setex(
```
[the remark this answers is not in the document]
Comment (amycommits) [reply api/app/Jobs/UpdateUserWeather.php:54]: Oh! I've worked on projects with Redis before, but I have never had to set it up from scratch. I will make these changes before requesting a re-review.

File: api/env_for_cursor.txt
Comment (amycommits) [reply api/env_for_cursor.txt:1]: Yeah, I definitely meant to delete this. Let's pretend that I committed it in order to ask your opinion on how those two new config variables should be stored. 

You're absolutely correct that API keys and other secrets should never be committed to Github.
